tokio/src/runtime/blocking/sharded_queue.rs

```diff
@@ -0,0 +1,285 @@
+//! A sharded concurrent queue for the blocking pool.
+//!
+//! This implementation distributes tasks across multiple shards to reduce
+//! lock contention when many threads are spawning blocking tasks concurrently.
+//! The push operations use per-shard locking, while notifications use a global
+//! condvar for simplicity.
+//!
+//! The queue adapts to the current concurrency level by using fewer shards
+//! when there are few threads, which improves cache locality and reduces
+//! lock contention on the active shards.
+
+use crate::loom::sync::{Condvar, Mutex};
+
+use std::collections::VecDeque;
+use std::sync::atomic::Ordering::{Acquire, Relaxed, Release};
+use std::sync::atomic::{AtomicBool, AtomicUsize};
+use std::time::Duration;
+
+use super::pool::Task;
+
+/// Number of shards. Must be a power of 2.
+const NUM_SHARDS: usize = 16;
+
+/// A single shard containing a queue protected by its own mutex.
+struct Shard {
+    /// The task queue for this shard.
+    queue: Mutex<VecDeque<Task>>,
```

Comment:
Compared to `std::sync::mpsc`, what advantages does `Mutex<VecDeque>` have in this scenario?

Reply:
Hmm, how would you imagine using mpsc? The overall design is to have multiple workers all pulling from a shared queue -- this just shards that queue so there's less contention, but multiple workers can still grab tasks (and multiple threads add them) concurrent.

If there was a mpmc module we could use that to pass tasks around and delete most of this (assuming it had good scalability)

I think if we wanted an alternate design, we could use something like crossbeam-queue's SegQueue.

Would you prefer I tried that?

Reply:
I think you mean the `mpsc::Receiver` doesn't allow threads to pull tasks from other shards. If so, I agree with you.